packages/client/src/tlog/verify/checkpoint.ts

```diff
@@ -0,0 +1,231 @@
+export function verifyCheckpoint(
+  entry: sigstore.VerifiableTransparencyLogEntry,
+  tlogs: sigstore.TransparencyLogInstance[]
+) {
+  const inclusionProof = entry.inclusionProof;
+  if (!inclusionProof) {
+    throw new VerificationError('tlog entry has no inclusion proof');
+  }
+
+  const envelope = inclusionProof?.checkpoint?.envelope;
+  if (!envelope) {
+    throw new VerificationError('tlog entry has no checkpoint');
+  }
+
+  // Filter tlog instances to just those which were valid at the time of the
+  // entry
+  const validTLogs = filterTLogInstances(tlogs, entry.integratedTime);
+
+  const signedNote = SignedNote.fromString(envelope);
+  const checkpoint = LogCheckpoint.fromString(signedNote.note);
+
+  // Verify that the signatures in the checkpoint are all valid
+  if (!signedNote.verify(validTLogs)) {
+    throw new VerificationError(
+      'inclusion proof contains invalid checkpoint signature'
+    );
+  }
+
+  // Verify that the root hash from the checkpoint matches the root hash in the
+  // inclusion proof
+  if (!crypto.bufferEqual(checkpoint.logHash, inclusionProof.rootHash)) {
+    throw new VerificationError(
+      'inclusion proof contains invalid root hash signature'
+    );
+  }
+}
+
+// SignedNote represents a signed note from a transparency log checkpoint. Consists
+// of a body (or note) and one more signatures calculated over the body. See
+// https://github.com/transparency-dev/formats/blob/main/log/README.md#signed-envelope
+class SignedNote {
+  readonly note: string;
+  readonly signatures: TLogSignature[];
+
+  constructor(note: string, signatures: TLogSignature[]) {
+    this.note = note;
+    this.signatures = signatures;
+  }
+
+  // Deserialize a SignedNote from a string
+  static fromString(envelope: string): SignedNote {
+    if (!envelope.includes(CHECKPOINT_SEPARATOR)) {
+      throw new VerificationError('malformed checkpoint: no separator');
+    }
+
+    // Split the note into the header and the data portions at the separator
+    const split = envelope.indexOf(CHECKPOINT_SEPARATOR);
+    const header = envelope.slice(0, split + 1);
+    const data = envelope.slice(split + CHECKPOINT_SEPARATOR.length);
+
+    // Find all the signature lines in the data portion
+    const matches = data.matchAll(SIGNATURE_REGEX);
+
+    // Parse each of the matched signature lines into the name and signature.
+    // The first four bytes of the signature are the key hint (should match the
+    // first four bytes of the log ID), and the rest is the signature itself.
+    const signatures = Array.from(matches, (match) => {
+      const [, name, signature] = match;
+      const sigBytes = Buffer.from(signature, 'base64');
+
+      if (sigBytes.length < 5) {
+        throw new VerificationError('malformed checkpoint: invalid signature');
+      }
+
+      return {
+        name,
+        keyHint: sigBytes.subarray(0, 4),
+        signature: sigBytes.subarray(4),
```

Comment:
Poking at the python implementation this jumped out: https://github.com/sigstore/sigstore-python/blob/bbf0ac83b18dda81aac1e472982507b91e34a8f8/sigstore/_internal/rekor/checkpoint.py#L165

Any idea why they seem to be encoding the sig as base64? Or is it already double encoded or something?

Reply:
oh think I found how this works, looks like it's decoded here again: https://github.com/sigstore/sigstore-python/blob/main/sigstore/_internal/rekor/checkpoint.py#L184